druid-shell/src/scale.rs

```diff
@@ -0,0 +1,289 @@
+            dpi_y: SCALE_TARGET_DPI * scale_y,
+            scale_x,
+            scale_y,
+            size_pt: Size::ZERO,
+            size_px: Size::ZERO,
+        }
+    }
+
+    /// Set the size of the scaled area in pixels.
+    ///
+    /// This updates the internal state and returns the same size in points.
+    pub fn set_size_px<T: Into<Size>>(&mut self, size: T) -> Size {
+        let size = size.into();
+        self.size_px = size;
+        self.size_pt = self.px_to_pt_size(size);
+        self.size_pt
+    }
+
+    /// Set the size of the scaled area in pixels via conversion from points.
+    ///
+    /// This updates the internal state and returns the same size in pixels.
+    ///
+    /// The calculated size in pixels is rounded away from zero to integers.
+    /// That means that the scaled area size in points isn't always the same
+    /// as the `size` given to this method. To find out the new size in points use [`size_pt`].
+    ///
+    /// [`size_pt`]: #method.size_pt
+    pub fn set_size_pt<T: Into<Size>>(&mut self, size: T) -> Size {
+        let size = size.into();
+        self.size_px = self.pt_to_px_size(size).expand();
+        self.size_pt = self.px_to_pt_size(self.size_px);
+        self.size_px
+    }
+
+    /// Returns the x axis platform DPI associated with this `Scale`.
+    #[inline]
+    pub fn dpi_x(&self) -> f64 {
+        self.dpi_x
+    }
+
+    /// Returns the y axis platform DPI associated with this `Scale`.
+    #[inline]
+    pub fn dpi_y(&self) -> f64 {
+        self.dpi_y
+    }
+
+    /// Returns `true` if the specified DPI is approximately equal to the `Scale` DPI.
+    #[inline]
+    pub fn dpi_approx_eq(&self, dpi_x: f64, dpi_y: f64) -> bool {
+        self.dpi_x.approx_eq(dpi_x, (f64::EPSILON, 2))
+            && self.dpi_y.approx_eq(dpi_y, (f64::EPSILON, 2))
+    }
+
+    /// Returns the x axis scale factor.
+    #[inline]
+    pub fn scale_x(&self) -> f64 {
+        self.scale_x
+    }
+
+    /// Returns the y axis scale factor.
+    #[inline]
+    pub fn scale_y(&self) -> f64 {
+        self.scale_y
+    }
+
+    /// Returns the scaled area size in points.
+    #[inline]
+    pub fn size_pt(&self) -> Size {
+        self.size_pt
+    }
+
+    /// Returns the scaled area size in pixels.
+    #[inline]
+    pub fn size_px(&self) -> Size {
+        self.size_px
+    }
+
+    /// Converts from points into pixels, using the x axis scale factor.
+    #[inline]
+    pub fn pt_to_px_x<T: Into<f64>>(&self, x: T) -> f64 {
+        x.into() * self.scale_x
+    }
+
+    /// Converts from points into pixels, using the y axis scale factor.
+    #[inline]
+    pub fn pt_to_px_y<T: Into<f64>>(&self, y: T) -> f64 {
+        y.into() * self.scale_y
+    }
+
+    /// Converts from points into pixels,
+    /// using the x axis scale factor for `x` and the y axis scale factor for `y`.
+    #[inline]
+    pub fn pt_to_px_xy<T: Into<f64>>(&self, x: T, y: T) -> (f64, f64) {
```

Comment:
this is a very annoying suggestion from me, _but_: if we had a trait `Scaleable`, which we implemented for our geometry types, we could just have,

```rust
trait Scaleable {
    fn to_display_point(&self, scale: &Scale) -> Self;
    fn to_pixel(&self, scale: &Scale) -> Self;
}

impl Scale {
    fn to_display_point<T: Scaleable>(&self, geom: &T) -> T {
        geom.to_display_point(self)
    }
    // etc
}
```

This would at least reduce the overall API surface?